charts/kong/templates/_helpers.tpl

```diff
@@ -354,156 +320,175 @@ Use the Pod security context defined in Values or set the UID by default
+      {{- $smtpPassword := include "secretkeyref" (dict "name" .Values.enterprise.smtp.auth.smtp_password_secret "key" "smtp_password") -}}
+      {{- $_ := set $autoEnv "KONG_SMTP_PASSWORD" $smtpPassword -}}
+    {{- end }}
+  {{- else }}
+    {{- $_ := set $autoEnv "KONG_SMTP_MOCK" "on" -}}
+  {{- end }}
+
+  {{- $lic := include "secretkeyref" (dict "name" .Values.enterprise.license_secret "key" "license") -}}
+  {{- $_ := set $autoEnv "KONG_LICENSE_DATA" $lic -}}
+
+{{- end }} {{/* End of the Enterprise settings block */}}
+
+{{- $_ := set $autoEnv "KONG_NGINX_HTTP_INCLUDE" "/kong/servers.conf" -}}
+
+{{- if .Values.postgresql.enabled }}
+  {{- $_ := set $autoEnv "KONG_PG_HOST" (include "kong.postgresql.fullname" .) -}}
+  {{- $_ := set $autoEnv "KONG_PG_PORT" .Values.postgresql.service.port -}}
+  {{- $pgPassword := include "secretkeyref" (dict "name" (include "kong.postgresql.fullname" .) "key" "postgresql-password") -}}
+  {{- $_ := set $autoEnv "KONG_PG_PASSWORD" $pgPassword -}}
 {{- end }}
-{{- if .Values.enterprise.smtp.enabled }}
-- name: KONG_PORTAL_EMAILS_FROM
-  value: {{ .Values.enterprise.smtp.portal_emails_from }}
-- name: KONG_PORTAL_EMAILS_REPLY_TO
-  value: {{ .Values.enterprise.smtp.portal_emails_reply_to }}
-- name: KONG_ADMIN_EMAILS_FROM
-  value: {{ .Values.enterprise.smtp.admin_emails_from }}
-- name: KONG_ADMIN_EMAILS_REPLY_TO
-  value: {{ .Values.enterprise.smtp.admin_emails_reply_to }}
-- name: KONG_SMTP_HOST
-  value: {{ .Values.enterprise.smtp.smtp_host }}
-- name: KONG_SMTP_PORT
-  value: {{ .Values.enterprise.smtp.smtp_port | quote }}
-- name: KONG_SMTP_STARTTLS
-  value: {{ .Values.enterprise.smtp.smtp_starttls | quote }}
-{{- if .Values.enterprise.smtp.auth.smtp_username }}
-- name: KONG_SMTP_USERNAME
-  value: {{ .Values.enterprise.smtp.auth.smtp_username }}
-- name: KONG_SMTP_PASSWORD
-  valueFrom:
-    secretKeyRef:
-      name: {{ .Values.enterprise.smtp.auth.smtp_password_secret }}
-      key: smtp_password
+
+{{- if (and (not .Values.ingressController.enabled) (eq .Values.env.database "off")) }}
+  {{- $_ := set $autoEnv "KONG_DECLARATIVE_CONFIG" "/kong_dbless/kong.yml" -}}
 {{- end }}
+
+{{- $_ := set $autoEnv "KONG_PLUGINS" (include "kong.plugins" .) -}}
+
+{{/*
+    ====== USER-SET ENVIRONMENT VARIABLES ======
+*/}}
+
+{{- $userEnv := dict -}}
+{{- range $key, $val := .Values.env }}
+  {{- $upper := upper $key -}}
+  {{- $var := printf "KONG_%s" $upper -}}
+  {{- $_ := set $userEnv $var $val -}}
+{{- end -}}
+
+{{/*
+      ====== MERGE AND RENDER ENV BLOCK ======
+*/}}
+
+{{- $completeEnv := mergeOverwrite $autoEnv $userEnv -}}
+
+{{- range keys $completeEnv }}
+{{- $val := pluck . $completeEnv | first -}}
+{{- $valueType := printf "%T" $val -}}
+{{ if eq $valueType "map[string]interface {}" }}
+- name: {{ . }}
+{{ toYaml $val | indent 2 -}}
+{{- else if eq $valueType "string" }}
+{{- if regexMatch "valueFrom" $val }}
+- name: {{ . }}
+{{ $val | indent 2 }}
```

Comment:
Can we indent this set of for loops and ifs for a little bit readability?
There are no conventions but having something in this one would help long-term.

Reply:
Maybe. It's easy to do this for the blocks above because nothing there gets rendered and all whitespace is chomped, but for the render block, both whitespace and indentation matter, and getting them right is tricky. I can do something like this easily:

```
{{- range keys $completeEnv }}
  {{- $val := pluck . $completeEnv | first -}} 
  {{- $valueType := printf "%T" $val -}} 
  {{ if eq $valueType "map[string]interface {}" }}
- name: {{ . }}
{{ toYaml $val | indent 2 -}} 
  {{- else if eq $valueType "string" }}
    {{- if regexMatch "valueFrom" $val }}
- name: {{ . }}
{{ $val | indent 2 }}
    {{- else }}
- name: {{ . }}
  value: {{ $val | quote }}
    {{- end }}
  {{- else }}
- name: {{ . }}
  value: {{ $val | quote }}
  {{- end }}
{{- end -}} 
```
But IMO this is actually less readable than the flat version, since you can't easily tell whether indentation is something we render or if it's to indicate the depth within control flow blocks. With the flat version, any indentation is something that's rendered in the template.

There may be a way to structure this such that we can freely indent both the control flow stuff and rendered content within so that they both reflect the control flow depth, and indentation for rendered content is controlled independently. A bunch of `{{- printf whatever | indent X}}` statements is probably a viable approach.

However, Helm makes this very annoying to debug when you get it wrong: any invalid YAML breaks the entire template process, and produces nothing more something like this very unhelpful error:
```
Error: YAML parse error on kong/templates/migrations.yaml: error converting YAML to JSON: yaml: line 38: could not find expected ':'
```
The line number there is in the rendered invalid YAML, which Helm refuses to show you. Their official guidance is to comment out the section you think you broke, but it's not always obvious which (sometimes you chomped a newline you shouldn't have and broke a section other than the one you're modifying) and sometimes you can't easily comment everything (with nested templates, you have to comment out content in all of them).

So, ultimately, yes, but not yet, and contingent on me or someone else writing a Helm patch that lets it just dump the offending YAML when it fails to convert to JSON.

Reply:
Templating YAML with Go templates was such a good idea.

Let's keep it as is. This was more about readability but as as you very well summarized, correctness would be hard to guarantee and it's well worth to sacrifice readability here.
Thank you for the detailed explanations, as always!